File: leetcode-cn/12.整数转罗马数字.py
#
# @lc app=leetcode.cn id=12 lang=python3
#
# [12] 整数转罗马数字
#


# @lc code=start
class Solution:
    def intToRoman(self, num: int) -> str:
        ans = []

        values = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
        value_char = {
            1000: 'M',
            900: 'CM',
            500: 'D',
            400: 'CD',
            100: 'C',
            90: 'XC',
            50: 'L',
            40: 'XL',
            10: 'X',
            9: 'IX',
            5: 'V',
            4: 'IV',
            1: 'I'
        }

        while num != 0:
            for value in values:
                if value <= num:
                    ans.append(value_char[value])
                    num -= value
                    break

        return ''.join(ans)


# @lc code=end

File: leetcode-cn/C++/12.整数转罗马数字.cpp
/*
 * @lc app=leetcode.cn id=12 lang=cpp
 *
 * [12] 整数转罗马数字
 */

// @lc code=start
#include <string>
using namespace std;
class Solution {
public:
    string intToRoman(int num)
    {
        string strs[] = { "M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I" };
        int nums[] = { 1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1 };

        string ans = "";
        for (int i = 0; i < 13; i++) {
            if (num == 0) {
                break;
            }
            while (num >= nums[i]) {
                ans += strs[i];
                num -= nums[i];
            }
        }
        return ans;
    }
};
// @lc code=end
